File: delivery-info-service/src/test/java/com/potatorider/deliveryinfoservice/publisher/DeliveryMessagePublisherImplTests.java
package com.potatorider.deliveryinfoservice.publisher;

import static com.potatorider.domain.DeliveryStatus.ACCEPT;
import static com.potatorider.domain.DeliveryStatus.REQUEST;

import static org.mockito.Mockito.any;
import static org.mockito.Mockito.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import com.potatorider.deliveryinfoservice.domain.DeliverySteps;
import com.potatorider.domain.Delivery;
import com.potatorider.publihser.DeliveryMessagePublisherImpl;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.amqp.core.AmqpTemplate;

import reactor.test.StepVerifier;

@ExtendWith(MockitoExtension.class)
public class DeliveryMessagePublisherImplTests {

    private final String shopExchange = "messageQueue.exchange.shop";
    private final String agencyExchange = "messageQueue.exchange.agency";
    DeliveryMessagePublisherImpl deliveryPublisher;
    AmqpTemplate amqpTemplate;

    @BeforeEach
    void setUp() {
        amqpTemplate = mock(AmqpTemplate.class);
        deliveryPublisher = new DeliveryMessagePublisherImpl(amqpTemplate);
    }

    @Test
    void send_add_delivery_event() {
        // Arrange
        var delivery = DeliverySteps.makeValidDeliveryWithDeliveryStatus(REQUEST);

        // Act
        var result = deliveryPublisher.sendAddDeliveryEvent(delivery);

        // Assert
        StepVerifier.create(result).expectNext(delivery).verifyComplete();
        verify(amqpTemplate, times(1))
                .convertAndSend(eq(shopExchange), any(String.class), any(Delivery.class));
    }

    @Test
    void send_set_rider_event() {
        // Arrange
        var delivery = DeliverySteps.makeValidDeliveryWithDeliveryStatus(ACCEPT);

        // Act
        var result = deliveryPublisher.sendSetRiderEvent(delivery);

        // Assert
        StepVerifier.create(result).expectNext(delivery).verifyComplete();
        verify(amqpTemplate, times(1))
                .convertAndSend(eq(agencyExchange), any(String.class), any(Delivery.class));
    }
}

File: delivery-info-service/src/test/java/com/potatorider/deliveryinfoservice/publisher/DeliveryMessagePublisherImplTests.kt
package com.potatorider.deliveryinfoservice.publisher

import com.potatorider.deliveryinfoservice.domain.makeValidDeliveryWithDeliveryStatus
import com.potatorider.domain.Delivery
import com.potatorider.domain.DeliveryStatus
import com.potatorider.publihser.DeliveryMessagePublisherImpl
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mockito
import org.mockito.Mockito.*
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.amqp.core.AmqpTemplate
import reactor.test.StepVerifier

@ExtendWith(MockitoExtension::class)
class DeliveryMessagePublisherImplTests {
    private val shopExchange = "messageQueue.exchange.shop"
    private val agencyExchange = "messageQueue.exchange.agency"
    private lateinit var deliveryPublisher: DeliveryMessagePublisherImpl
    private lateinit var amqpTemplate: AmqpTemplate

    @BeforeEach
    fun setUp() {
        amqpTemplate = mock(AmqpTemplate::class.java)
        val actualDeliveryPublisher = DeliveryMessagePublisherImpl(amqpTemplate)
        deliveryPublisher = spy(actualDeliveryPublisher)
    }

    @Test
    fun send_add_delivery_event() {
        // Arrange
        val delivery = makeValidDeliveryWithDeliveryStatus(DeliveryStatus.REQUEST)

        // Act
        val result = deliveryPublisher.sendAddDeliveryEvent(delivery)

        // Assert
        StepVerifier.create(result).expectNext(delivery).verifyComplete()
        verify(amqpTemplate, times(1))
            .convertAndSend(
                eq(shopExchange), any(
                    String::class.java
                ), any(Delivery::class.java)
            )
    }

    @Test
    fun send_set_rider_event() {
        // Arrange
        val delivery = makeValidDeliveryWithDeliveryStatus(DeliveryStatus.ACCEPT)

        // Act
        val result = deliveryPublisher.sendSetRiderEvent(delivery)

        // Assert
        StepVerifier.create(result).expectNext(delivery).verifyComplete()
        verify(amqpTemplate, times(1))
            .convertAndSend(
                eq(agencyExchange), any(
                    String::class.java
                ), any(Delivery::class.java)
            )
    }
}
